internal/flowcontrol/writebuffer.go

```diff
@@ -65,16 +66,20 @@ func (w *ResourceSliceWriteBuffer) PatchStatusAsync(ctx context.Context, ref *re
 	currentSlice := w.state[key]
 	for i, item := range currentSlice {
 		if *item.SlicedResource == *ref {
+			// last write wins
 			currentSlice[i].PatchFn = patchFn
 			return
 		}
 	}
 
+	w.insertionTime[key] = time.Now()
 	w.state[key] = append(currentSlice, &resourceSliceStatusUpdate{
 		SlicedResource: ref,
 		PatchFn:        patchFn,
 	})
-	w.queue.Add(key)
+	if w.queue.NumRequeues(key) == 0 {
```

Comment:
Could you explain the reasoning behind this change? How does the queue behave when calling Add on an item that's already in it?

Reply:
Calling `Add` while the same key is in-flight will cause it to be requeued after `Done` is called regardless of the rate limiter. `AddRateLimited` ultimately just calls `Add` but increments honors the rate limiter state.

In this case the first attempt should count against the rate limit and calls to PatchStatusAsync shouldn't extend or reset the rate limit. Then we increment the retry counter (in `processQueueItem`) before calling Done to back off any hot loops.